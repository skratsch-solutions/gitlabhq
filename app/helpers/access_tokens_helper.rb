# frozen_string_literal: true

module AccessTokensHelper
  include AccountsHelper
  include ApplicationHelper

  def scope_description(prefix)
    case prefix
    when :project_access_token
      [:doorkeeper, :project_access_token_scope_desc]
    when :group_access_token
      [:doorkeeper, :group_access_token_scope_desc]
    else
      [:doorkeeper, :scope_desc]
    end
  end

  def tokens_app_data
    {
      feed_token: {
        enabled: !Gitlab::CurrentSettings.disable_feed_token,
        token: current_user.feed_token,
        reset_path: reset_feed_token_profile_path
      },
      incoming_email_token: {
        enabled: incoming_email_token_enabled?,
        token: current_user.enabled_incoming_email_token,
        reset_path: reset_incoming_email_token_profile_path
      },
      static_object_token: {
        enabled: static_objects_external_storage_enabled?,
        token: current_user.enabled_static_object_token,
        reset_path: reset_static_object_token_profile_path
      }
    }.to_json
  end

  def expires_at_field_data
    {
      min_date: 1.day.from_now.iso8601,
      max_date: max_date_allowed
    }
  end

  private

  def max_date_allowed
    return unless Gitlab::CurrentSettings.require_personal_access_token_expiry?

    ::PersonalAccessToken.max_expiration_lifetime_in_days.days.from_now.iso8601
  end
end

AccessTokensHelper.prepend_mod
