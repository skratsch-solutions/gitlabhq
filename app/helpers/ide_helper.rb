# frozen_string_literal: true

module IdeHelper
  # Overridden in EE
  def ide_data(project:, fork_info:, params:)
    base_data = {
      'use-new-web-ide' => use_new_web_ide?.to_s,
      'new-web-ide-help-page-path' => help_page_path('user/project/web_ide/index.md'),
      'sign-in-path' => new_session_path(current_user),
      'sign-out-path' => destroy_user_session_path,
      'user-preferences-path' => profile_preferences_path
    }.merge(use_new_web_ide? ? new_ide_data(project: project) : legacy_ide_data(project: project))

    return base_data unless project

    base_data.merge(
      'fork-info' => fork_info&.to_json,
      'branch-name' => params[:branch],
      'file-path' => params[:path],
      'merge-request' => params[:merge_request_id]
    )
  end

  def show_web_ide_oauth_callback_mismatch_callout?
    return false unless ::WebIde::DefaultOauthApplication.feature_enabled?(current_user)

    callback_urls = ::WebIde::DefaultOauthApplication.oauth_application_callback_urls
    callback_url_domains = callback_urls.map { |url| URI.parse(url).origin }
    callback_url_domains.any? && callback_url_domains.exclude?(request.base_url)
  end

  def web_ide_oauth_application_id
    ::WebIde::DefaultOauthApplication.oauth_application_id
  end

  def use_new_web_ide?
    Feature.enabled?(:vscode_web_ide, current_user)
  end

  private

  def new_ide_fonts
    {
      fallback_font_family: 'monospace',
      font_faces: [{
        family: 'GitLab Mono',
        display: 'block',
        src: [{
          url: font_url('gitlab-mono/GitLabMono.woff2'),
          format: 'woff2'
        }]
      }, {
        family: 'GitLab Mono',
        display: 'block',
        style: 'italic',
        src: [{
          url: font_url('gitlab-mono/GitLabMono-Italic.woff2'),
          format: 'woff2'
        }]
      }]
    }
  end

  def new_ide_code_suggestions_data
    {}
  end

  def new_ide_oauth_data
    return {} unless ::WebIde::DefaultOauthApplication.feature_enabled?(current_user)
    return {} unless ::WebIde::DefaultOauthApplication.oauth_application

    client_id = ::WebIde::DefaultOauthApplication.oauth_application.uid
    callback_urls = ::WebIde::DefaultOauthApplication.oauth_application_callback_urls

    {
      'client-id' => client_id,
      'callback-urls' => callback_urls
    }
  end

  def new_ide_data(project:)
    extensions_gallery_settings = WebIde::ExtensionsMarketplace.webide_extensions_gallery_settings(user: current_user)
    settings_context_hash = WebIde::SettingsSync.settings_context_hash(
      extensions_gallery_settings: extensions_gallery_settings
    )

    {
      'project-path' => project&.path_with_namespace,
      'csp-nonce' => content_security_policy_nonce,
      'editor-font' => new_ide_fonts.to_json,
      'extensions-gallery-settings' => extensions_gallery_settings.to_json,
      'settings-context-hash' => settings_context_hash
    }.merge(new_ide_code_suggestions_data).merge(new_ide_oauth_data)
  end

  def legacy_ide_data(project:)
    {
      'empty-state-svg-path' => image_path('illustrations/empty-state/empty-variables-md.svg'),
      'no-changes-state-svg-path' => image_path('illustrations/status/status-nothing-sm.svg'),
      'committed-state-svg-path' => image_path('illustrations/rocket-launch-md.svg'),
      'pipelines-empty-state-svg-path': image_path('illustrations/empty-state/empty-pipeline-md.svg'),
      'switch-editor-svg-path': image_path('illustrations/rocket-launch-md.svg'),
      'ci-help-page-path' => help_page_path('ci/quick_start/index.md'),
      'web-ide-help-page-path' => help_page_path('user/project/web_ide/index.md'),
      'render-whitespace-in-code': current_user.render_whitespace_in_code.to_s,
      'default-branch' => project && project.default_branch,
      'project' => convert_to_project_entity_json(project),
      'preview-markdown-path' => project && preview_markdown_path(project),
      'web-terminal-svg-path' => image_path('illustrations/empty-state/empty-cloud-md.svg'),
      'web-terminal-help-path' => help_page_path('user/project/web_ide/index.md'),
      'web-terminal-config-help-path' => help_page_path('user/project/web_ide/index.md'),
      'web-terminal-runners-help-path' => help_page_path('user/project/web_ide/index.md')
    }
  end

  def convert_to_project_entity_json(project)
    return unless project

    API::Entities::Project.represent(project, current_user: current_user).to_json
  end

  def has_dismissed_ide_environments_callout?
    current_user.dismissed_callout?(feature_name: 'web_ide_ci_environments_guidance')
  end
end
