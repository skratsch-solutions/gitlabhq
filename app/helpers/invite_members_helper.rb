# frozen_string_literal: true

module InviteMembersHelper
  include Gitlab::Utils::StrongMemoize

  def can_invite_members_for_project?(project)
    # do not use the can_admin_project_member? helper here due to structure of the view and how membership_locked?
    # is leveraged for inviting groups
    can?(current_user, :admin_project_member, project)
  end

  def invite_accepted_notice(member)
    format(
      _('You have been granted access to the %{source_name} %{source_type} with the following role: %{role_name}.'),
      source_name: member.source.name,
      source_type: member.source.model_name.singular,
      role_name: member.present.human_access
    )
  end

  # Overridden in EE
  def common_invite_group_modal_data(source, member_class)
    {
      id: source.id,
      root_id: source.root_ancestor.id,
      name: source.name,
      default_access_level: Gitlab::Access::GUEST,
      invalid_groups: source.related_group_ids,
      help_link: help_page_url('user/permissions.md'),
      is_project: source.is_a?(Project).to_s,
      access_levels: member_class.permissible_access_level_roles(current_user, source).to_json,
      full_path: source.full_path
    }.merge(group_select_data(source))
  end

  # Overridden in EE
  def common_invite_modal_dataset(source)
    {
      id: source.id,
      root_id: source.root_ancestor&.id,
      name: source.name,
      default_access_level: Gitlab::Access::GUEST,
      full_path: source.full_path
    }
  end

  private

  def group_select_data(source)
    if source.root_ancestor.prevent_sharing_groups_outside_hierarchy
      { groups_filter: 'descendant_groups', parent_id: source.root_ancestor.id }
    else
      {}
    end
  end

  # Overridden in EE
  def users_filter_data(group)
    {}
  end
end

InviteMembersHelper.prepend_mod_with('InviteMembersHelper')
