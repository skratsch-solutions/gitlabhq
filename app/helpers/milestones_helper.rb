# frozen_string_literal: true

module MilestonesHelper
  def milestone_header_class(primary, issuables)
    header_color = milestone_header_color(primary: primary)
    header_border = milestone_header_border(issuables)

    "#{header_color} #{header_border} gl-flex"
  end

  def milestone_counter_class(primary)
    primary ? 'gl-text-white' : 'gl-text-subtle'
  end

  private

  def milestone_header_color(primary: false)
    return '' unless primary

    'gl-bg-blue-500 gl-text-white'
  end

  def milestone_header_border(issuables)
    issuables.empty? ? 'gl-border-b-0 gl-rounded-base' : ''
  end
end
