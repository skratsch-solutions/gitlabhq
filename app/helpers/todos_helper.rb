# frozen_string_literal: true

module TodosHelper
  def todos_pending_count
    @todos_pending_count ||= current_user.todos_pending_count
  end

  def todos_done_count
    @todos_done_count ||= current_user.todos_done_count
  end

  def todo_action_name(todo) # rubocop:disable Metrics/CyclomaticComplexity -- Will be removed/refactored as part of https://gitlab.com/gitlab-org/gitlab/-/issues/464069
    case todo.action
    when Todo::ASSIGNED then todo.self_added? ? _('assigned') : _('assigned you')
    when Todo::REVIEW_REQUESTED then s_('Todos|requested a review')
    when Todo::MENTIONED, Todo::DIRECTLY_ADDRESSED then format(
      s_("Todos|mentioned %{who}"), who: todo_action_subject(todo)
    )
    when Todo::BUILD_FAILED then s_('Todos|The pipeline failed')
    when Todo::MARKED then s_('Todos|added a to-do item')
    when Todo::APPROVAL_REQUIRED then format(
      s_("Todos|set %{who} as an approver"), who: todo_action_subject(todo)
    )
    when Todo::UNMERGEABLE then s_('Todos|Could not merge')
    when Todo::MERGE_TRAIN_REMOVED then s_("Todos|Removed from Merge Train")
    when Todo::MEMBER_ACCESS_REQUESTED then format(
      s_("Todos|has requested access to %{what} %{which}"), what: _(todo.member_access_type), which: _(todo.target.name)
    )
    when Todo::REVIEW_SUBMITTED then s_('Todos|reviewed your merge request')
    when Todo::OKR_CHECKIN_REQUESTED then format(
      s_("Todos|requested an OKR update for %{what}"), what: todo.target.title
    )
    when Todo::SSH_KEY_EXPIRED then s_('Todos|Your SSH key has expired')
    when Todo::SSH_KEY_EXPIRING_SOON then s_('Todos|Your SSH key is expiring soon')
    end
  end

  def todo_self_addressing(todo)
    case todo.action
    when Todo::ASSIGNED then _('to yourself')
    when Todo::REVIEW_REQUESTED then _('from yourself')
    end
  end

  def todo_target_name(todo)
    return todo.target_reference unless todo.for_commit?

    content_tag(:span, todo.target_reference, class: 'commit-sha')
  end

  def todo_target_title(todo)
    # Design To Dos' filenames are displayed in `#todo_target_name` (see `Design#to_reference`),
    # so to avoid displaying duplicate filenames in the To Do list for designs,
    # we return an empty string here.
    return "" if todo.target.blank? || todo.for_design? || todo.member_access_requested?

    todo.target.title.to_s
  end

  def todo_parent_path(todo)
    if todo.resource_parent.is_a?(Group)
      todo.resource_parent.name
    else
      # Note: Some todos (like for expired SSH keys) are neither related to a project nor a group.
      return unless todo.project.present?

      title = content_tag(:span, todo.project.name, class: 'project-name')
      namespace = content_tag(:span, "#{todo.project.namespace.human_name} / ", class: 'namespace-name')

      title.prepend(namespace) if todo.project.namespace

      title
    end
  end

  def todo_target_aria_label(todo)
    target_type = if todo.for_design?
                    _('Design')
                  elsif todo.for_alert?
                    _('Alert')
                  elsif todo.member_access_requested?
                    _('Group')
                  elsif todo.for_issue_or_work_item?
                    IntegrationsHelper.integration_issue_type(todo.target.issue_type)
                  else
                    IntegrationsHelper.integration_todo_target_type(todo.target_type)
                  end

    "#{target_type} #{todo_target_name(todo)}"
  end

  def todo_target_path(todo)
    return unless todo.target.present?

    path_options = todo_target_path_options(todo)

    if todo.for_commit?
      project_commit_path(todo.project, todo.target, path_options)
    elsif todo.for_design?
      todos_design_path(todo, path_options)
    elsif todo.for_alert?
      details_project_alert_management_path(todo.project, todo.target)
    elsif todo.for_issue_or_work_item?
      path_options[:only_path] = true
      Gitlab::UrlBuilder.build(todo.target, **path_options)
    elsif todo.for_ssh_key?
      user_settings_ssh_key_path(todo.target)
    elsif todo.member_access_requested?
      todo.access_request_url(only_path: true)
    else
      path = [todo.resource_parent, todo.target]

      path.unshift(:pipelines) if todo.build_failed?

      polymorphic_path(path, path_options)
    end
  end

  def todo_target_path_options(todo)
    { anchor: todo_target_path_anchor(todo) }
  end

  def todo_target_path_anchor(todo)
    dom_id(todo.note) if todo.note.present?
  end

  def todo_target_state_pill(todo)
    return unless show_todo_state?(todo)

    state = todo.target.state.to_s
    raw_state_to_i18n = {
      "closed" => _('Closed'),
      "merged" => _('Merged'),
      "resolved" => _('Resolved')
    }

    case todo.target
    when MergeRequest
      case state
      when 'closed'
        variant = 'danger'
      when 'merged'
        variant = 'info'
      end
    when Issue
      variant = 'info' if state == 'closed'
    when AlertManagement::Alert
      variant = 'info' if state == 'resolved'
    else
      variant = 'info'
    end

    content_tag(:span, class: 'todo-target-state') do
      gl_badge_tag(raw_state_to_i18n[state] || state.capitalize, { variant: variant })
    end
  end

  def todos_filter_params
    {
      state: params[:state].presence,
      project_id: params[:project_id],
      author_id: params[:author_id],
      type: params[:type],
      action_id: params[:action_id]
    }.compact
  end

  def todos_filter_empty?
    todos_filter_params.values.none?
  end

  def todos_has_filtered_results?
    params[:group_id] || params[:project_id] || params[:author_id] || params[:type] || params[:action_id]
  end

  def no_todos_messages
    [
      s_('Todos|Good job! Looks like you don\'t have anything left on your To-Do List'),
      s_('Todos|Isn\'t an empty To-Do List beautiful?'),
      s_('Todos|Give yourself a pat on the back!'),
      s_('Todos|Nothing left to do. High five!'),
      s_('Todos|Henceforth, you shall be known as "To-Do Destroyer"')
    ]
  end

  def todos_filter_path(options = {})
    without = options.delete(:without)

    options = todos_filter_params.merge(options)

    if without.present?
      without.each do |key|
        options.delete(key)
      end
    end

    "#{request.path}?#{options.to_param}"
  end

  def todo_actions_options
    [
      { id: '', text: s_('Todos|Any Action') },
      { id: Todo::ASSIGNED, text: s_('Todos|Assigned') },
      { id: Todo::REVIEW_REQUESTED, text: s_('Todos|Review requested') },
      { id: Todo::MENTIONED, text: s_('Todos|Mentioned') },
      { id: Todo::MARKED, text: s_('Todos|Added') },
      { id: Todo::BUILD_FAILED, text: s_('Todos|Pipelines') },
      { id: Todo::MEMBER_ACCESS_REQUESTED, text: s_('Todos|Member access requested') },
      { id: Todo::SSH_KEY_EXPIRED, text: s_('Todos|SSH key expired') },
      { id: Todo::SSH_KEY_EXPIRING_SOON, text: s_('Todos|SSH key expiring soon') }
    ]
  end

  def todo_types_options
    [
      { id: '', text: s_('Todos|Any Type') },
      { id: 'Issue', text: s_('Todos|Issue') },
      { id: 'MergeRequest', text: s_('Todos|Merge request') },
      { id: 'DesignManagement::Design', text: s_('Todos|Design') },
      { id: 'AlertManagement::Alert', text: s_('Todos|Alert') },
      { id: 'Key', text: s_('Todos|SSH key') }
    ]
  end

  def todo_actions_dropdown_label(selected_action_id, default_action)
    selected_action = todo_actions_options.find { |action| action[:id] == selected_action_id.to_i }
    selected_action ? selected_action[:text] : default_action
  end

  def todo_types_dropdown_label(selected_type, default_type)
    selected_type = todo_types_options.find { |type| type[:id] == selected_type && type[:id] != '' }
    selected_type ? selected_type[:text] : default_type
  end

  def todo_due_date(todo)
    return unless todo.target.try(:due_date)

    is_due_today = todo.target.due_date.today?
    is_overdue = todo.target.overdue?
    css_class =
      if is_due_today
        'gl-text-warning'
      elsif is_overdue
        'gl-text-danger'
      else
        ''
      end

    due_date =
      if is_due_today
        _("today")
      else
        l(todo.target.due_date, format: Date::DATE_FORMATS[:medium])
      end

    content = content_tag(:span, class: css_class) do
      format(s_("Todos|Due %{due_date}"), due_date: due_date)
    end

    "#{content} &middot;".html_safe
  end

  def todo_author_display?(todo)
    !todo.build_failed? && !todo.unmergeable? && !todo.for_ssh_key?
  end

  def todo_groups_requiring_saml_reauth(_todos)
    []
  end

  private

  def todos_design_path(todo, path_options)
    design = todo.target

    designs_project_issue_path(
      todo.resource_parent,
      design.issue,
      path_options.merge(
        vueroute: design.filename
      )
    )
  end

  def todo_action_subject(todo)
    todo.self_added? ? s_('Todos|yourself') : _('you')
  end

  def show_todo_state?(todo)
    case todo.target
    when MergeRequest, Issue
      %w[closed merged].include?(todo.target.state)
    when AlertManagement::Alert
      %i[resolved].include?(todo.target.state)
    else
      false
    end
  end
end

TodosHelper.prepend_mod_with('TodosHelper')
