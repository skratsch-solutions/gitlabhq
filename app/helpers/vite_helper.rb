# frozen_string_literal: true

module ViteHelper
  def vite_enabled?
    # vite is not production ready yet
    return false if Rails.env.production?

    Gitlab::Utils.to_boolean(ViteRuby.env['VITE_ENABLED'], default: false)
  end

  def vite_hmr_websocket_url
    ViteRuby.env['VITE_HMR_WS_URL']
  end

  def vite_hmr_http_url
    ViteRuby.env['VITE_HMR_HTTP_URL']
  end

  def vite_page_entrypoint_paths
    action = case controller.action_name
             when 'create' then 'new'
             when 'update' then 'edit'
             else controller.action_name
             end

    parts = (controller.controller_path.split('/') << action)

    parts.map
         .with_index { |part, idx| "pages.#{(parts[0, idx] << part).join('.')}.js" }
  end

  def universal_stylesheet_link_tag(path, **options)
    return stylesheet_link_tag(path, **options) unless vite_enabled?

    if Rails.env.test? && config.asset_host
      # Link directly to Vite server when running tests because for unit and integration tests, there
      # won't be a Rails server to proxy these requests to the Vite server.
      options[:host] = URI::HTTP.build(host: ViteRuby.config.host, port: ViteRuby.config.port).to_s
    end

    options[:extname] = false

    stylesheet_link_tag(
      ViteRuby.instance.manifest.path_for("stylesheets/styles.#{path}.scss", type: :stylesheet),
      **options
    )
  end

  def universal_path_to_stylesheet(path, **options)
    return ActionController::Base.helpers.stylesheet_path(path, **options) unless vite_enabled?

    ViteRuby.instance.manifest.path_for("stylesheets/styles.#{path}.scss", **options)
  end
end
