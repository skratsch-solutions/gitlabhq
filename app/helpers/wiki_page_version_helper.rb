# frozen_string_literal: true

module WikiPageVersionHelper
  include SafeFormatHelper

  def wiki_page_version_author_url(wiki_page_version)
    user = wiki_page_version.author
    user.nil? ? "mailto:#{wiki_page_version.author_email}" : Gitlab::UrlBuilder.build(user)
  end

  def wiki_page_version_author_avatar(wiki_page_version)
    image_tag(avatar_icon_for_email(wiki_page_version.author_email, 24), class: "avatar s24 float-none !gl-mr-0")
  end
end
