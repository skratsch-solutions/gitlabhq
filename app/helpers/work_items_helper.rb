# frozen_string_literal: true

module WorkItemsHelper
  def work_items_show_data(resource_parent, current_user)
    group = resource_parent.is_a?(Group) ? resource_parent : resource_parent.group

    {
      autocomplete_award_emojis_path: autocomplete_award_emojis_path,
      can_admin_label: can?(current_user, :admin_label, resource_parent).to_s,
      full_path: resource_parent.full_path,
      group_path: group&.full_path,
      issues_list_path:
        resource_parent.is_a?(Group) ? issues_group_path(resource_parent) : project_issues_path(resource_parent),
      labels_manage_path:
        resource_parent.is_a?(Group) ? group_labels_path(resource_parent) : project_labels_path(resource_parent),
      register_path: new_user_registration_path(redirect_to_referer: 'yes'),
      sign_in_path: new_session_path(:user, redirect_to_referer: 'yes'),
      new_comment_template_paths: new_comment_template_paths(group).to_json,
      report_abuse_path: add_category_abuse_reports_path,
      default_branch: resource_parent.is_a?(Project) ? resource_parent.default_branch_or_main : nil,
      initial_sort: current_user&.user_preference&.issues_sort,
      is_signed_in: current_user.present?.to_s,
      show_new_issue_link: can?(current_user, :create_work_item, group).to_s,
      can_create_projects: can?(current_user, :create_projects, group).to_s,
      new_project_path: new_project_path(namespace_id: group&.id),
      group_id: group&.id
    }
  end

  # overriden in EE
  def add_work_item_show_breadcrumb(resource_parent, _iid)
    path = resource_parent.is_a?(Group) ? issues_group_path(resource_parent) : project_issues_path(resource_parent)

    add_to_breadcrumbs(_('Issues'), path)
  end

  def work_items_list_data(group, current_user)
    {
      autocomplete_award_emojis_path: autocomplete_award_emojis_path,
      full_path: group.full_path,
      initial_sort: current_user&.user_preference&.issues_sort,
      is_signed_in: current_user.present?.to_s,
      show_new_issue_link: can?(current_user, :create_work_item, group).to_s,
      issues_list_path: issues_group_path(group),
      report_abuse_path: add_category_abuse_reports_path,
      labels_manage_path: group_labels_path(group),
      can_admin_label: can?(current_user, :admin_label, group).to_s
    }
  end
end
