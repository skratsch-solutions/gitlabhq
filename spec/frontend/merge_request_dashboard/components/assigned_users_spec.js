import { mountExtended } from 'helpers/vue_test_utils_helper';
import AssignedUsers from '~/merge_request_dashboard/components/assigned_users.vue';

let wrapper;
let glTooltipDirectiveMock;

const createMockUsers = () => [
  {
    id: 'gid://gitlab/user/1',
    webUrl: '/root',
    name: 'Admin',
    avatarUrl: '/root',
  },
  {
    id: 'gid://gitlab/user/2',
    webUrl: '/root',
    name: 'Admin',
    avatarUrl: '/root',
    mergeRequestInteraction: {
      reviewState: 'REQUESTED_CHANGES',
    },
  },
];

function createComponent({
  users = createMockUsers(),
  type = 'ASSIGNEES',
  newListsEnabled = false,
} = {}) {
  glTooltipDirectiveMock = jest.fn();

  wrapper = mountExtended(AssignedUsers, {
    directives: {
      GlTooltip: glTooltipDirectiveMock,
    },
    provide: {
      newListsEnabled,
    },
    propsData: {
      type,
      users,
    },
  });
}

describe('Merge request dashboard assigned users component', () => {
  const findAllUsers = () => wrapper.findAllByTestId('assigned-user');
  const findReviewStateIcon = () => wrapper.findByTestId('review-state-icon');
  const findCurrentUserIcon = () => wrapper.findByTestId('current-user');

  beforeEach(() => {
    window.gon = { current_user_id: 1 };
  });

  afterEach(() => {
    window.gon = {};
  });

  it('renders user avatars', () => {
    createComponent();

    expect(findAllUsers()).toHaveLength(2);
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('current user avatar', () => {
    it('renders icon for current user', () => {
      createComponent();

      expect(findCurrentUserIcon().exists()).toBe(true);
    });

    it('adds this is you text to tooltip', () => {
      createComponent();

      expect(glTooltipDirectiveMock.mock.calls[1][1].value).toBe(
        '<strong>This is you.</strong><br />Assigned to Admin',
      );
    });

    it('renders current user last', () => {
      createComponent();

      expect(findAllUsers().at(1).find('[data-testid="current-user"]').exists()).toBe(true);
    });

    describe('when newListsEnabled is true', () => {
      it('renders icon for current user', () => {
        createComponent({ newListsEnabled: true });

        expect(findCurrentUserIcon().exists()).toBe(false);
      });
    });
  });

  describe('as reviewer list', () => {
    it('renders review state icon', () => {
      createComponent({ type: 'REVIEWER' });

      expect(findReviewStateIcon().exists()).toBe(true);
      expect(findReviewStateIcon().html()).toMatchSnapshot();
    });

    it.each`
      state                  | title
      ${'REQUESTED_CHANGES'} | ${'Admin requested changes'}
      ${'APPROVED'}          | ${'Approved by Admin'}
      ${'REVIEWED'}          | ${'Admin left feedback'}
      ${'UNREVIEWED'}        | ${'Review requested from Admin'}
    `('sets title as $title for review state $state', ({ state, title }) => {
      createComponent({
        type: 'REVIEWER',
        users: [
          {
            id: 'gid://gitlab/user/2',
            webUrl: '/root',
            name: 'Admin',
            avatarUrl: '/root',
            mergeRequestInteraction: {
              reviewState: state,
            },
          },
        ],
      });

      expect(glTooltipDirectiveMock.mock.calls[0][1].value).toBe(title);
    });
  });
});
