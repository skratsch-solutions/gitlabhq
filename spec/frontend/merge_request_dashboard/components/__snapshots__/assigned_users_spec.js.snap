// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merge request dashboard assigned users component as reviewer list renders review state icon 1`] = `
<span
  class="-gl-bottom-2 -gl-right-2 gl-absolute gl-bg-status-danger gl-flex gl-h-5 gl-items-center gl-justify-center gl-p-1 gl-rounded-full gl-w-5"
  data-testid="review-state-icon"
>
  <svg
    aria-hidden="true"
    class="gl-block gl-fill-current gl-fill-status-danger gl-icon s12"
    data-testid="error-icon"
    role="img"
  >
    <use
      href="file-mock#error"
    />
  </svg>
</span>
`;

exports[`Merge request dashboard assigned users component renders user avatars 1`] = `
<div
  class="gl-flex gl-justify-center mr-users-list"
>
  <div
    class="gl-avatars-inline gl-avatars-inline-lg"
  >
    <div
      class="gl-avatars-inline-child"
    >
      <a
        class="gl-avatar-link gl-link gl-relative"
        data-testid="assigned-user"
        href="/root"
        target="blank"
      >
        <img
          alt="avatar"
          class="!gl-bg-white gl-avatar gl-avatar-circle gl-avatar-s32"
          src="/root"
        />
        <span
          class="-gl-bottom-2 -gl-right-2 gl-absolute gl-bg-status-danger gl-flex gl-h-5 gl-items-center gl-justify-center gl-p-1 gl-rounded-full gl-w-5"
          data-testid="review-state-icon"
        >
          <svg
            aria-hidden="true"
            class="gl-block gl-fill-current gl-fill-status-danger gl-icon s12"
            data-testid="error-icon"
            role="img"
          >
            <use
              href="file-mock#error"
            />
          </svg>
        </span>
      </a>
    </div>
    <div
      class="gl-avatars-inline-child"
    >
      <a
        class="gl-avatar-link gl-link gl-relative"
        data-testid="assigned-user"
        href="/root"
        target="blank"
      >
        <img
          alt="avatar"
          class="!gl-bg-white gl-avatar gl-avatar-circle gl-avatar-s32"
          src="/root"
        />
        <span
          class="-gl-left-2 -gl-top-2 gl-absolute gl-bg-blue-500 gl-flex gl-h-5 gl-items-center gl-justify-center gl-p-1 gl-rounded-full gl-text-white gl-w-5"
          data-testid="current-user"
        >
          <svg
            aria-hidden="true"
            class="gl-block gl-fill-current gl-icon s12"
            data-testid="user-icon"
            role="img"
          >
            <use
              href="file-mock#user"
            />
          </svg>
        </span>
      </a>
    </div>
  </div>
</div>
`;
